crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1037,13 +1173,99 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
         // We handled the case that the closing quotes are on their own line
         // above (the last line is empty except for whitespace). If they are on
         // the same line as content, we don't insert a line break.
-        if !self.is_last {
+        if !line.is_last {
             hard_line_break().fmt(self.f)?;
         }
 
         Ok(())
     }
 
+    /// Given a sequence of lines from a code snippet, format them and return
+    /// the formatted code as a sequence of owned docstring lines.
+    ///
+    /// This routine generally only returns an error when the recursive call
+    /// to the formatter itself returns a `FormatError`. In all other cases
+    /// (for example, if the code snippet is invalid Python or even if the
+    /// resulting reformatted code snippet is invalid Python), then `Ok(None)`
+    /// is returned. In this case, callers should assume that a reformatted
+    /// code snippet is unavailable and bail out of trying to format it.
+    ///
+    /// Currently, when the above cases happen and `Ok(None)` is returned, the
+    /// routine is silent about it. So from the user's perspective, this will
+    /// fail silently. Ideally, this would at least emit a warning message,
+    /// but at time of writing, it wasn't clear to me how to best do that.
+    fn format(
+        &mut self,
+        code: &[CodeExampleLine<'src>],
+    ) -> FormatResult<Option<Vec<DocstringLine<'static>>>> {
+        use ruff_python_parser::AsMode;
+
+        let offset = code
+            .get(0)
+            .expect("code blob must be non-empty")
+            .original
+            .offset;
+        let last_line_is_last = code
+            .last()
+            .expect("code blob must be non-empty")
+            .original
+            .is_last;
+        let codeblob = code
+            .iter()
+            .map(|line| &*line.code)
+            .collect::<Vec<&str>>()
+            .join("\n");
```

Comment:
How about line endings inside of multiline strings? Ideally, we would preserve these. Although we may deliberately decide not to and encourage users to instead use explicit line endings. But changing the line endings inside of strings is theoretically a semantic change.

Reply:
I think Charlie had a similar comment above. The thinking here is that while I'm using `\n` to assemble a code snippet, that snippet is then fed into the formatter and the line endings used correspond to the user's setting. There is a test for example that this writes out CRLF line endings in code snippets when the user has configured CRLF line endings.

Did I understand you concern correctly here? I feel like I might be missing it.

Reply:
Also, there are places in the docstring handling (that existed before my changes) that seem to assume line feed terminators? For example:

https://github.com/astral-sh/ruff/blob/2ade84a29de99eafbac23ef4c6dc9cdf1101e014/crates/ruff_python_formatter/src/expression/string/docstring.rs#L237-L238

Reply:
Yeah, I think the problem is not new due to your changes. It only became evident that this might now be a problem when thinking about code snippets in docstrings (it shouldn't be a problem for non-code text). 

```python
a = """A multiline string
with windows line endings
"""

b = c
```
<img width="859" alt="Screenshot 2023-11-29 at 07 23 49" src="https://github.com/astral-sh/ruff/assets/1203881/f6eb79e4-0844-486a-bb95-046de837fd6d">

My concern is (was) that changing the newlines **inside** of the multiline string could be a semantic change. I tried to read through the Python documentation and only found:

> In triple-quoted literals, unescaped newlines and quotes are allowed (and are retained), except that three unescaped quotes in a row terminate the literal. (A “quote” is the character used to open the literal, i.e. either ' or ".) [source](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals)

What I understand from the spec is that newlines are retained. Although running the above in a `print` on my mac normalizes the newlines to `\n`. 

I then tested what the about for

```python
a = b"""A multiline string
with windows line endings
"""

from binascii import hexlify

print(hexlify(a))
```

and it seems python normalizes the newlines even for binary strings to `\n`. So the retain only means that the newlines are preserved but not in the form they're present in the source code.

So I guess this is not a problem after all (and ruff and Black both normalize newlines inside multiline strings)